main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1316,30 +1351,44 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   /**
     * Returns the implicit type parameters constructed from the given formal parameters and return type.
     */
-  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): List[NamedAst.TypeParam] = {
+  private def getImplicitTypeParams(fparams: List[WeededAst.FormalParam], returnType: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
+    // We use `freeVarsWithKind` to infer the kind for each free type variable in the signature.
+    // Then we ensure each occurrence of the same name maps to the same kind.
+    // Finally, we create a type param for each name.
+
     // Compute the type variables that occur in the formal parameters.
-    val typeVarsArgs = fparams.foldLeft(Set.empty[String]) {
-      case (acc, WeededAst.FormalParam(_, _, None, _)) => acc
-      case (acc, WeededAst.FormalParam(_, _, Some(tpe), _)) =>
-        freeVars(tpe).foldLeft(acc) {
-          case (innerAcc, ident) => innerAcc + ident.name
-        }
+    val typeVarsWithKindArgs = fparams.flatMap {
+      case WeededAst.FormalParam(_, _, Some(tpe), _) => freeVarsWithKind(tpe)
+      case WeededAst.FormalParam(_, _, None, _) => List.empty
     }
 
-    // Compute the type variables that occur in the return type.
-    val typeVarsReturnType = freeVars(returnType).map(_.name)
+    // Compute the type variables that occur in the return type
+    val typeVarsWithKindReturnType = freeVarsWithKind(returnType)
 
     // Compute the set of type variables.
-    val typeVars = typeVarsArgs ++ typeVarsReturnType
+    val typeVarsWithKind = typeVarsWithKindReturnType ::: typeVarsWithKindArgs
+
+    // create a map of name -> (ident, kind), ensuring all kinds for a given name match
+    val kindPerName = foldRight(typeVarsWithKind)(Map.empty[String, (Name.Ident, Kind)].toSuccess) {
```

Comment:
OK; I guess the map has to have this form because we don't have a "usable" equality on `Name.Ident`.

Reply:
Looks much better, btw